pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java

```diff
@@ -594,6 +594,10 @@ protected BrokerResponse handleRequest(long requestId, String query, @Nullable S
       long routingStartTimeNs = System.nanoTime();
       Map<ServerInstance, List<String>> offlineRoutingTable = null;
       Map<ServerInstance, List<String>> realtimeRoutingTable = null;
+      Map<ServerInstance, List<String>> optionalOfflineRoutingTable = null;
+      Map<ServerInstance, List<String>> optionalRealtimeRoutingTable = null;
+      Set<ServerInstance> offlineServerInstances = null;
+      Set<ServerInstance> realtimeServerInstances = null;
```

Comment:
It is much simpler if we change routing table to `Map<ServerInstance, Pair<List<String>, List<String>>` where the first list is mandatory segments and the second list is option segments. We may also change the `RoutingTable._serverInstanceToSegmentsMap` type